--- a/source/backend/plugin/CarlaPluginVST3.cpp
+++ b/source/backend/plugin/CarlaPluginVST3.cpp
@@ -410,12 +410,17 @@
         CARLA_SAFE_ASSERT_RETURN(num_bytes > 0, V3_INVALID_ARG);
         CARLA_SAFE_ASSERT_RETURN(stream->canWrite, V3_INVALID_ARG);
 
-        void* const newbuffer = std::realloc(stream->buffer, stream->size + num_bytes);
-        CARLA_SAFE_ASSERT_RETURN(newbuffer != nullptr, V3_NOMEM);
+        if (stream->readPos + num_bytes > stream->size)
+        {
+            void* const newbuffer = std::realloc(stream->buffer, stream->readPos + num_bytes);
+            CARLA_SAFE_ASSERT_RETURN(newbuffer != nullptr, V3_NOMEM);
+
+            stream->buffer = newbuffer;
+            stream->size = stream->readPos + num_bytes;
+        }
 
-        std::memcpy(static_cast<uint8_t*>(newbuffer) + stream->size, buffer, num_bytes);
-        stream->buffer = newbuffer;
-        stream->size += num_bytes;
+        std::memcpy(static_cast<uint8_t*>(stream->buffer) + stream->readPos, buffer, num_bytes);
+        stream->readPos += num_bytes;
 
         // this is nasty, some plugins do not care about incomplete writes!
         if (bytes_read != nullptr)
@@ -427,7 +432,7 @@
     static v3_result V3_API seek(void* const self, const int64_t pos, const int32_t seek_mode, int64_t* const result)
     {
         carla_v3_bstream* const stream = *static_cast<carla_v3_bstream**>(self);
-        CARLA_SAFE_ASSERT_RETURN(stream->canRead, V3_INVALID_ARG);
+        CARLA_SAFE_ASSERT_RETURN(stream->canRead || stream->canWrite, V3_INVALID_ARG);
 
         switch (seek_mode)
         {
@@ -457,7 +462,7 @@
     {
         carla_v3_bstream* const stream = *static_cast<carla_v3_bstream**>(self);
         CARLA_SAFE_ASSERT_RETURN(pos != nullptr, V3_INVALID_ARG);
-        CARLA_SAFE_ASSERT_RETURN(stream->canRead, V3_INVALID_ARG);
+        CARLA_SAFE_ASSERT_RETURN(stream->canRead || stream->canWrite, V3_INVALID_ARG);
 
         *pos = stream->readPos;
         return V3_OK;
@@ -4041,6 +4046,24 @@
                 v3_cpp_obj(connController)->connect(connController, connComponent);
             }
 
+            {
+                carla_v3_bstream stream;
+                carla_v3_bstream* const streamPtr = &stream;
+                v3_bstream** const v3stream = (v3_bstream**)&streamPtr;
+
+                stream.canWrite = true;
+
+                if (v3_cpp_obj(component)->get_state(component, v3stream) == V3_OK)
+                {
+                    stream.canRead = true;
+                    stream.canWrite = false;
+                    stream.readPos = 0;
+                    v3_cpp_obj(controller)->set_component_state(controller, v3stream);
+                }
+
+                std::free(stream.buffer);
+            }
+
             // get midi mapping interface
             if (v3_cpp_obj_query_interface(component, v3_midi_mapping_iid, &midiMapping) != V3_OK)
             {
